Fix convert_escape_to_noescape non-dominance errors

aschwaighofer opened the pull request.

aschwaighofer committed:
SILGen: Don't postpone the cleanup of the convertEscapeToNoEscape inside an bind optional

This would violate dominance since the convertEscapeToNoEscape
instruction does not dominate the postpone point.

rdar://38124009

aschwaighofer committed:
Ensure lifetimes of optional convert_escape_to_noescape closures

rdar://38124009

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swift-ci Please test source compatibility

atrick commented on lib/SILGen/SILGenBuilder.cpp:
```diff
@@ -209,7 +209,8 @@ ManagedValue SILGenBuilder::createConvertFunction(SILLocation loc,
                                                           ManagedValue fn,
-                                                          SILType resultTy) {
+                                                          SILType resultTy,
+                                                          bool dontPostponeToNoEscapeCleanup) {
```
80-col (git format)

atrick reviewed (commented):
~~I don't see any change in functionality. Just default arguments. And I don't see any new CHECK lines in the test~~ (only saw the first commit first time around).

atrick commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -2090,8 +2090,11 @@ RValue RValueEmitter::visitFunctionConversionExpr(FunctionConversionExpr *e,
+    result = SGF.emitTransformedValue(e, result, srcTy, destTy, SGFContext(),
+                                      dontPostponeToNoEscapeCleanup);
+  }
 
```
Ignore my previous comment. I'm not sure why this didn't show up in my first review. It's not clear why you want to use a double negative everywhere instead of just `postponeNoEscapeCleanup = true`.

atrick commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -2952,6 +2952,114 @@ static bool lowerRawSILOperations(SILFunction &Fn) {
+///   We will insert a copy_value of the original %closure before the two
+///   diamonds. And a destroy of %closure at the last destroy of
+///   %convertOptionalBlock.
+static bool fixupConvertEscapeToNoEscapeLifetime(SILFunction &Fn) {
```
Ok, but *why* are you doing this transformation? Please explain in the comment why the SIL input to DI is invalid.

aschwaighofer commented:
This is a plaster until I have time to remove PostponedCleanup in favor of a full implementation (the mark_initalized way your are eluding to) of handling the postponement in DI.  I want to get rid of postponement in SILGen.

slavapestov commented:
Instead of DI, can the long term fix introduce a new SIL pass? This is not really related to DI.

aschwaighofer committed:
Disable definite-init-convert-to-escape.swift test case on linux

aschwaighofer commented:
Sure.

aschwaighofer commented:
@swift-ci Please test linux

aschwaighofer commented:
@swift-ci Please smoke test os x

aschwaighofer committed:
SILGen: Remove variable with negated name:

aschwaighofer committed:
Add a convert_escape_to_noescape [not_guaranteed] variant

aschwaighofer committed:
Use the convert_escape_to_noescape [not_guaranteed] instruction

aschwaighofer committed:
Rewrite DI to always guaranteed the lifetime of the convert_escape_to_noescape [not_guaranteed] operand

Either through a peephole or we keep the closure operand alive to the end of the function

rdar://38124009

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
@swift-ci Please test source compatibility

aschwaighofer commented:
The new patchset adds patches that make sure we ensure the lifetime of the escaping closure: either through the peephole or by keeping the closure alive to the end of the function if the peephole fails.

aschwaighofer commented:
There should be documentation other than in my head that 'convert_escape_to_noescape [not_guaranteed]' means that a pass must be run after SILGen that ensures the lifetime of the escaping closure extends beyond the last use of the trivial closure result.

aschwaighofer committed:
Remove hack

aschwaighofer commented on lib/Serialization/SerializeSIL.cpp:
```diff
@@ -579,10 +580,18 @@ void SILSerializer::writeOneTypeOneOperandLayout(SILInstructionKind valueKind,
 /// Write an instruction that looks exactly like a conversion: all
 /// important information is encoded in the operand and the result type.
-void SILSerializer::writeConversionLikeInstruction(const SingleValueInstruction *I) {
+__attribute__((noinline))
```
This is going away in a follow-up.

aschwaighofer commented:
Latest source compatibility run: https://ci.swift.org/view/Pull%20Request/job/swift-PR-source-compat-suite/970/

aschwaighofer committed:
Fix logic error in fixupConvertEscapeToNoEscapeLifetime

aschwaighofer committed:
Really fix the logic error in fixupConvertEscapeToNoEscapeLifetime

aschwaighofer commented:
@swift-ci Please smoke test
